### ISSUE TITLE
Making unwrap_in_result warn by default?

### ISSUE THREAD

**Issue by kornelski** [rust-lang/rust-clippy#16159]
### Description

There's a lint against use of `.unwrap()` inside functions returning `Result`:

https://rust-lang.github.io/rust-clippy/master/index.html?search=UNWRAP_IN_RESULT#unwrap_in_result

Could it be enabled by default?

While `.unwrap()` could be [used intentionally in many places](https://burntsushi.net/unwrap/), IMHO mixing of `Result` and `.unwrap()` in the same function could more likely to be a sign of an oversight or bad style: copypaste from another panicking function, or leftover from before the function was refactored to return `Result`, a quick hack that hasn't been cleaned up, or a novice mistake from lack of understanding the distinction between panics and errors.

Panics from `Result`-returning functions can be surprising to the callers (e.g. the caller may `match` exhaustively on the `Err`, but fail to realize that there's another failure case to handle differently). Although lints like this one can't guarantee that functions won't unexpectedly panic, it may be a good idea to discourage that, especially from a potentially-not-carefully-thought-through `.unwrap()`.

The lint docs currently say "this can cause false positives in functions that handle both recoverable and non recoverable errors", which is true, but I don't think that's a big problem. The lint flags only `.unwrap()` and `.expect()`, but doesn't fire on `assert!`, `unreachable!`, and `panic!`, so there are still ways of handling unrecoverable situations without triggering this lint. 

If false positives are still too problematic for `unwrap_in_result` to warn by default, this lint (or its copy) could be narrowed down to fire only on `.unwrap()` (assuming `.expect()` is more intentional and helpful), skip `#[cfg(test)]`/`#[test]` code, and/or skip functions marked with `#[track_caller]` (such attribute explicitly acknowledges the function may panic, and it may be caller's fault).

### Version

```text

```

### Additional Labels

@rustbot label +C-question

**Comment by y21** [comment]
I'm personally fairly strongly opposed to this. I *mostly* agree with the linked post, and while I do agree that sometimes people incorrectly use `unwrap` for certain errors that ought to be propagated or handled properly, I don't think a linter should make such a general decision without context, and I don't think a function having `Result` as its return type has any relation to whether any error that happens in the function is a bug or not.

Blindly following the lint's suggestion of "consider using the `?` operator" (which I think is a possible outcome given that many beginners use clippy to learn new patterns) makes very little sense in many instances and can IMHO even *worsen* the situation.
It delays the bug and obfuscates the bug the further you propagate it away from where the broken invariant is first detected, and if exposed from a library there is nothing that a user can really usefully do with a "bug" error kind (and as burntsushi says in the post, leaks implementation details).

It's possible to rewrite the unwrap using a match and `panic!` (essentially inlining it), but that doesn't really seem at all different or useful to me.

Ultimately, this lint in its design is trying to discourage the intended way of using `unwrap()` and `expect()` , which is already a very common pattern used throughout the ecosystem and even [inside](https://github.com/rust-lang/rust/blob/b84478a1c477756cd3e1974eda867a6bb31e8902/library/std/src/os/xous/ffi.rs#L24) [the](https://github.com/rust-lang/rust/blob/b84478a1c477756cd3e1974eda867a6bb31e8902/library/std/src/sys/path/uefi.rs#L67) [standard library](https://github.com/rust-lang/rust/blob/b84478a1c477756cd3e1974eda867a6bb31e8902/library/std/src/sync/mpmc/zero.rs#L144), so this is not something I'd like to see from a lint that's enabled by default.

**Comment by kornelski** [comment]
Opinions on this differ. There was even a proposal to [deprecate `.unwrap()` entirely](https://internals.rust-lang.org/t/re-opening-deprecating-option-unwrap-and-result-unwrap/23734).

> It's possible to rewrite the unwrap using a match and panic! (essentially inlining it), but that doesn't really seem at all different or useful to me.

I think there is a substantial difference between use of `.unwrap()` and more specific alternatives like `assert!`, or even `.expect()`.

This is because the intent behind an `.unwrap()` in code is unclear:

* It could have been added by a novice user, copying some example code or just following other compiler suggestions, without fully realizing the implications. It could be a poor pattern like `if val.is_some() { let val = val.unwrap(); }`. This is exactly the kind of not-ideal code that Clippy should be teaching about!
* It could have been added by an experienced user who understands it exactly, but done as a quick stopgap to make the code run right away, and has slipped through code reviews or was forgotten and never replaced with a dedicated `Error` case or a better handling/fallback.
* It could have been meant to be exactly like this, but given the two other possibilities, the intent is not clear to people reviewing the code, so it's always going to raise eyebrows. Additionally, message from `.unwrap()` itself may be ambiguous and unhelpful. It's especially poor when called on `None`.
* It could be used with expressions that never fail (`str.split().next().unwrap()`). Admittedly, this is a relatively common case and will be a false positive. Maybe Rust should get some solution to this? (`unreachable!()` in a method form?) Maybe the lint can exclude certain patterns?

So even changing `.unwrap()` to `.expect("why this is here")` is an improvement (this lint could be split to make this distinction). It's self-documenting, and more helpful when it triggers. For checking function inputs, `assert!` and/or `#[track_caller]` communicates the intent better. So even if the goal is to trigger a panic on bad inputs/state inside a `Result`-returning function, it could be done better than a plain `.unwrap()`.

**Comment by RalfJung** [comment]
I am also strongly opposed. I use `unwrap` (and other forms of panics) regularly, including in `Result`-returning functions, to check internal invariants. It would make no sense at all to turn those into `Err`. This would become just yet another lint on the already-too-long list of clippy lints that I always end up disabling, moving the needle of "is clippy worth the effort" ever so slightly in the wrong direction.

**Comment by kornelski** [comment]
Could there be any minimal subset of this lint that you'd find acceptable? 

Are there any clues that could help minimize false positives on the intentional invariant-asserting `.unwrap()`s? I've already suggested allowing `.expect()`, and excluding `#[test]`/`#[track_caller]`. It could also be skipped if the doccomments mention panics/unwinding/invariants. Maybe even allow `.unwrap()` if there's a comment on that line or statement?

Maybe std could provide `.unwrap`-equivalent methods under other names to make it easy to mark specific invariants and distinguish from random not-thought-through unwraps? e.g. `.this_cant_ever_fail()`/`.i_already_checked_that()`/`mutex.lock_or_panic_if_poisoned()` (to be bikeshedded)?

**Comment by ojeda** [comment]
I also do not agree that `-> Result` always implies the function should not have panics (or just `.unwrap()`s in particular). It is true that in some cases that can happen, but it is generallyorthogonal. So making it a default warning means getting some false positives by default.

At least for certain projects, an alternative would be to require `// PANIC: ...` comments to explain why a panic call was written/needed, similar to the `// SAFETY: ...` lints: https://github.com/rust-lang/rust-clippy/issues/15895. This should give pause to developers before adding an `.unwrap()` call and solves the mentioned "ambiguity problem" (e.g. whether it was added as a "quick hack" or it was indeed intentional).

**Comment by RalfJung** [comment]
> Maybe std could provide .unwrap-equivalent methods under other names to make it easy to mark specific invariants and distinguish from random not-thought-through unwraps? e.g. .this_cant_ever_fail()/.i_already_checked_that()/mutex.lock_or_panic_if_poisoned() (to be bikeshedded)?

If there's some way other than `unwrap` to express "it is an invariant of the system that this must be `Some`", then 🤷 I guess I could switch all my code to that. But it shouldn't be overly verbose, i..e not much longer than `assert`.

The lint could also make a difference between `Result::unwrap` and `Option::unwrap`. That said, a common use of `unwrap` for me is `try_into().unwrap()`, and those (a) definitely should not be `Err`, they express logic bugs, and (b) do involve `Result::unwrap`. I wish we had a less verbose way of converting between integer types while enforcing a bounds check, but this is what we got.

That said, as Miguel said it does seem odd to tie this to whether the surrounding function has a `Result` return type or not. This is orthogonal. Just because a function has expected errors doesn't change how unexpected errors should be treated. And if the codebase abuses `unwrap` to handle expected errors, that's a problem everywhere, not just in places that return `Result`.
